src/runtime-rs/src/shim/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+    OciSpec(#[source] oci_spec::OciSpecError),
+    #[error("failed to extract oci spec information")]
+    OciSpecInfo(#[source] virtcontainers::oci::OciSpecInfoError),
+    #[error("failed to open file {1}")]
+    OpenFile(#[source] std::io::Error, std::path::PathBuf),
+    #[error("failed to read file {1}")]
+    ReadFile(#[source] std::io::Error, std::path::PathBuf),
+    #[error("failed to write file {1}")]
+    WriteFile(#[source] std::io::Error, std::path::PathBuf),
+    #[error("failed to get self exec")]
+    SelfExec(#[source] std::io::Error),
+    #[error("failed to bind socket at {1}")]
+    BindSocket(#[source] std::io::Error, String),
+    #[error("failed to spawn child")]
+    SpawnChild(#[source] std::io::Error),
+    #[error("failed to clean container {0}")]
+    CleanUp(String),
+    #[error("failed to get env variable")]
+    EnvVar(#[source] std::env::VarError),
+    #[error("failed to parse server fd {0}")]
+    ServerFd(String),
+    #[error("failed to start ttrpc server")]
+    StartServer(#[source] ttrpc::error::Error),
+    #[error("failed to wait ttrpc server when {0}")]
+    WaitServer(String),
+    #[error("failed to set logger")]
+    Logger,
+    #[error("error from config")]
+    Config(#[source] config::ConfigError),
+    #[error("error from sandbox")]
+    Sandbox(#[source] virtcontainers::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+const SOCKET_ROOT: &str = "/run/containerd";
+
+const ENV_SOCKET_FD: &str = "BIND_FD";
+
+/// Received commandline arguments or environment arguments.
+///
+/// For defail information, please refer to the
+/// [shim v2 spec](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md).
+#[derive(Debug, Default, Clone)]
+pub struct ShimArgs {
+    /// the id of the container
+    pub id: String,
+    /// the namespace for the container
+    pub namespace: String,
+    /// the address of the containerd's main socket
+    pub address: String,
+    /// the binary path to publish events back to containerd
+    pub publish_binary: String,
+    /// the path to the bundle to delete
+    pub bundle: String,
+    /// Whether or not to enable debug
+    pub debug: bool,
+}
+
+impl ShimArgs {
+    /// Check the shim argument object is vaild or not.
+    ///
+    /// The id, namespace, address and publish_binary are mandatory for START, RUN and DELETE.
+    /// And bundle is mandatory for DELETE.
+    pub fn validate(&mut self, is_delete: bool) -> Result<()> {
+        if self.id.is_empty()
+            || self.namespace.is_empty()
+            || self.address.is_empty()
+            || self.publish_binary.is_empty()
+        {
+            return Err(Error::InvalidArgument);
+        }
+        if is_delete && self.bundle.is_empty() {
+            return Err(Error::InvalidArgument);
+        }
+
+        if Self::is_component_dangerous(&self.id) || Self::is_component_dangerous(&self.namespace) {
+            return Err(Error::InvalidArgument);
```

Comment:
Is there any way to add more detail to the returned error to make it clear which condition failed?

Reply:
I am not sure about this. Adding more detail would lead to the code too complicated